File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    count = 0
    tw = 0
    sec = 1
    q = deque()

    while True:
        if count >= len(truck_weights):
            sec = max(sec, q[0][1] + bridge_length)
            q.popleft()
            if not q:
                break
        elif not q:
            q.append((truck_weights[count], sec))
            tw += truck_weights[count]
            sec += 1
            count += 1
        else:
            if tw + truck_weights[count] > weight or len(q) >= bridge_length:
                tmp = q[0][0]
                sec = max(sec, q[0][1] + bridge_length)
                tw -= tmp
                q.popleft()
            else:
                q.append((truck_weights[count], sec))
                tw += truck_weights[count]
                count += 1
                sec += 1

    return sec

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    int answer = 0;
    int count = 0;
    int bt = 0;
    int tw = 0;
    int sec = 1;
    queue<pair<int, int>> q;
    
    while(true) {
        if (count >= truck_weights.size()) {
            sec = max(sec, q.front().second + bridge_length);
            q.pop();
            if (q.size() == 0) {
                break;
            }
        }
        else if (q.size() == 0) {
            q.push(make_pair(truck_weights[count], sec));
            tw += truck_weights[count];
            sec++;
            count++;
        }
        else {
            if (tw + truck_weights[count] > weight || q.size() >= bridge_length) {
                int tmp = q.front().first;
                sec = max(sec, q.front().second + bridge_length);
                tw -= tmp;
                q.pop();
            }
            else {
                q.push(make_pair(truck_weights[count], sec));
                tw += truck_weights[count];
                count++;
                sec++;
            }
        }
    }
    
    return sec;
}